src/Microsoft.AspNetCore.ResponseCompression/ResponseCompressionMiddleware.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.ResponseCompression
+{
+    public class ResponseCompressionMiddleware
+    {
+        private readonly RequestDelegate _next;
+
+        private readonly IResponseCompressionProvider[] _compressionProviders;
+
+        private readonly HashSet<string> _mimeTypes;
+
+        private readonly int _minimumSize;
+
+        /// <summary>
+        /// Initialize the Response Compression middleware.
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="options"></param>
+        public ResponseCompressionMiddleware(RequestDelegate next, IOptions<ResponseCompressionOptions> options)
+        {
+            if (options.Value.MinimumSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(options.Value.MinimumSize));
+            }
+            _minimumSize = options.Value.MinimumSize;
+
+            if (options.Value.MimeTypes == null)
+            {
+                throw new ArgumentNullException(nameof(options.Value.MimeTypes));
+            }
+            _mimeTypes = new HashSet<string>(options.Value.MimeTypes);
+
+            _next = next;
+
+            var providers = options.Value.Providers;
+            if (providers == null)
+            {
+                providers = new IResponseCompressionProvider[]
+                {
+                    new GzipResponseCompressionProvider(CompressionLevel.Fastest)
+                };
+            }
+            _compressionProviders = providers.ToArray();
+        }
+
+        /// <summary>
+        /// Invoke the middleware.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public async Task Invoke(HttpContext context)
+        {
+            IResponseCompressionProvider compressionProvider = SelectProvider(context.Request.Headers[HeaderNames.AcceptEncoding]);
+
+            if (compressionProvider == null)
+            {
+                await _next(context);
+                return;
+            }
+
+            var bodyStream = context.Response.Body;
+
+            using (var uncompressedStream = new MemoryStream())
+            {
+                context.Response.Body = uncompressedStream;
```

Comment:
Is there any special test to add, or I can assume that in any cases the MIME type is available when the first byte is written in the body stream?